Add nvbal testing of example notebooks

pgbarletta opened the pull request:
Hi!
Bit of a big PR but most of the changes on the notebooks are minor tweaks so **nbval** tests pass. Pinging @micaela-matta 

**nbval** runs notebooks and checks the output it gets against the output that's stored in the notebook file, both `stdout` and `stderr`. This makes tests quite demanding as a lot of non-errors can make a test fail, for example, the memory location that matplotlib prints along with the plot, which varies every time the cell is executed (more on this later), yet I think this is necessary for our case since the cells output is teaching/documentation content.

So, about the changes on the notebooks. These include:

- Rounding out raw float numbers so only the most significant digits are checked. Ex:
![image](https://user-images.githubusercontent.com/5693483/200179288-24e3b619-b366-45fa-a9ac-549bdd423aa6.png)
- Adding the `'nbval-ignore-output'` tag on cells that won't reliably pass tests. Jupyter notebooks support a `tags` field that lets the user add attributes to a cell. **nbval** will run but skip the checking of cells marked with this tag. This was a last-resource when there was no way of getting a regular output (see next bullet point for an example).
- Some cell splitting so the lines that give variable output can be ignored, while the rest can be checked. For example, these 2 lines of code used to be in 1 cell, but since the downloading process will give very different output each time the cell is executed, it was relocated into a previous cell, which is ignored by **nbval**.
![image](https://user-images.githubusercontent.com/5693483/200178969-369a3177-8123-46f0-b24a-47696d33b0fb.png)
If for any reason the download fails, then the creation of the `Universe` will fail as well so, in a sense, the download is still being checked.
- `reduced_dimensions/pca.ipynb` now only generates 3 PC's instead of 5, to save some time. The text was also changed accordingly.

Other things worth of notice:
- Added another workflow just for testing that only runs when there's a push or a PR onto `develop`. I thought this was best so we don't test everything each time someone fix a typo.
- **nbval** gives a warning, I already opened a [PR](https://github.com/computationalmodelling/nbval/pull/191) to fix it, hopefully it will eventually get some attention.
- Some notebooks output couldn't be checked since they give numeric output and rounding wasn't feasible without losing information. Ex., from `trajectory_similarity/clustering_ensemble_similarity.ipynb`: 
![image](https://user-images.githubusercontent.com/5693483/200180548-c1f0dcfc-56da-4858-bbef-4d0c04389ec0.png)
- I had to greatly expand the `environment.yml` to be able to run (almost) all notebooks.
- I added a `tests/sanitize_output.cfg`, this is a sanitization configuration file **nbval** takes in. Sanitization in this case means replacement of text that matches any of the regexs listed in the `tests/sanitize_output.cfg` file. This prevents testing failure due to outputs that vary with each execution without going as far as ignoring  the whole cell with the `'nbval-ignore-output'`
- Notebooks from `trajectory_similarity` take the most time.
- `volumetric/density_analysis.ipynb` is not being checked since Cell[18] hangs the instance. It's also the biggest offender when it comes to time consumption and file size (87MB).
- `polymers_and_membranes/hole2.ipynb` is not being checked since this requires HOLE.

Questions: 
- Is there an easy way to find out the minimum required version of MDA for each notebook or can I set all notebooks minimum version to `2.0.0`? With the usage of the `results` attribute that's the case for most of them anyway.
- Can we speed up notebooks from  `trajectory_similarity`? Perhaps getting some shorter trajectories?
- Do we want to test HOLE as well?
- Can we lighten up `volumetric/density_analysis.ipynb` so we can check it as well?

Remaing tasks:
- [ ]  Decide on `volumetric/density_analysis.ipynb`
- [ ] Decide on `polymers_and_membranes/hole2.ipynb`
- [ ]  On a future PR: add a **nbval** section on the Contributing guide with this info and more. 

Let me know of any required fixes, I'll squash everything at the end.

edit:
List of notebooks with at least 1 ignored cell
```
other/parmed_sim.ipynb
quickstart.ipynb
analysis/alignment_and_rms/rmsf.ipynb
analysis/reduced_dimensions/pca.ipynb
analysis/structure/site_specific_rdf.ipynb
analysis/volumetric/linear_density.ipynb (2 cells)
analysis/trajectory_similarity/clustering_ensemble_similarity.ipynb (2 cells)
analysis/trajectory_similarity/convergence.ipynb
analysis/trajectory_similarity/dimension_reduction_ensemble_similarity.ipynb (4 cells)
analysis/custom_parallel_analysis.ipynb (3 cells)
```

IAlibay commented on .github/workflows/gh-ci.yaml:
```diff
@@ -30,11 +30,14 @@ env:
-      CYTHON_TRACE_NOGIL: 1
-      MPLBACKEND: agg
+  tests:
+    runs-on: ${{ matrix.OS }}-latest
+    name: "examples-tests"
+    strategy:
+      fail-fast: False
+      matrix:
+        os: ['ubuntu']
+        python-version: [3.8]
```
Why a matrix if there's only one entry?

pgbarletta replied:
my bad, I did all of this before your latest commits and it seems `git rebase` just overwrote the file without showing any conflicts.

IAlibay commented:
> Added another workflow just for testing that only runs when there's a push onto develop. I thought this was best so we don't test everything each time someone fix a typo.

Not sure I agree with this, checking _after_ the PR is merged is problematic imho, we should know before merge that things need fixing.

IAlibay commented on .github/workflows/gh-ci.yaml:
```diff
@@ -30,11 +30,14 @@ env:
-      MPLBACKEND: agg
+  tests:
+    runs-on: ${{ matrix.OS }}-latest
+    name: "examples-tests"
```
Don't call both the same please.

pgbarletta replied:
Fixed.

IAlibay commented on .github/workflows/gh-ci.yaml:
```diff
@@ -30,11 +30,14 @@ env:
-    env:
-      CYTHON_TRACE_NOGIL: 1
-      MPLBACKEND: agg
+  tests:
```
please keep the old name - this is a build docs CI run

pgbarletta replied:
Fixed.

IAlibay commented on .github/workflows/gh-ci-tests.yaml:
```diff
@@ -0,0 +1,55 @@
+
+jobs:
+  tests:
+    runs-on: ${{ matrix.OS }}-latest
```
Same question - are we expecting a matrix here?

pgbarletta replied:
Fixed.

pgbarletta commented:
> > Added another workflow just for testing that only runs when there's a push onto develop. I thought this was best so we don't test everything each time someone fix a typo.
> 
> Not sure I agree with this, checking _after_ the PR is merged is problematic imho, we should know before merge that things need fixing.

My mistake, it also checks on PRs onto develop. Fixed the description now.

IAlibay commented on .github/workflows/gh-ci-tests.yaml:
```diff
@@ -0,0 +1,55 @@
+        run: |
+          mamba info -a
+          mamba list
+      - name: "test notebooks"
```
```suggestion
      - name: "test notebooks"

```

IAlibay commented on .github/workflows/gh-ci-tests.yaml:
```diff
@@ -0,0 +1,55 @@
+          mamba list
+      - name: "test notebooks"
+        run: |
+          pytest -v doc/source/examples/analysis/alignment_and_rms --nbval --current-env --sanitize-with tests/sanitize_output.cfg
```
Can we use a pytest.ini file instead?

IAlibay commented on .github/workflows/gh-ci-tests.yaml:
```diff
@@ -0,0 +1,55 @@
+  pull_request:
+    branches:
+      - develop
+
```
```suggestion

concurrency:
  # Probably overly cautious group naming.
  # Commits to develop/master will cancel each other, but PRs will only cancel
  # commits within the same PR
  group: "${{ github.ref }}-${{ github.head_ref }}-${{ github.workflow }}"
  cancel-in-progress: true

```
Add concurrency check

IAlibay commented on .github/workflows/gh-ci-tests.yaml:
```diff
@@ -0,0 +1,55 @@
+      - develop
+  pull_request:
+    branches:
+      - develop
```
```suggestion
      - develop
  schedule:
    # 3 am Tuesdays and Fridays
    - cron: "0 3 * * 2,5"
```
Add cron

IAlibay reviewed (changes requested):
Review solely on the new CI file - I'll try to do the rest over reviewnb

IAlibay commented on doc/source/examples/analysis/alignment_and_rms/aligning_trajectory.ipynb:
```diff
@@ -49,7 +49,7 @@
     {
```
<p>As above on the simtk import</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='3'/>

pgbarletta replied:
<p>Fixed.</p>

IAlibay commented on doc/source/examples/analysis/alignment_and_rms/aligning_trajectory.ipynb:
```diff
@@ -49,7 +49,7 @@
     {
```
<p>Probably will want to use a warning filter here too?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='7'/>

pgbarletta replied:
<p>Fixed.</p>

IAlibay commented on doc/source/examples/analysis/alignment_and_rms/aligning_trajectory.ipynb:
```diff
@@ -49,7 +49,7 @@
     {
```
<p>This is somewhat strange, and is rather inconsistent with everything else, why the change? Is it purely because running the notebooks creates a file? Can we not add it to a gitignore instead?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='23'/>

pgbarletta replied:
<p>Fixed, *.xyz files were already ignored.</p>

IAlibay commented on doc/source/examples/analysis/alignment_and_rms/aligning_trajectory_to_frame.ipynb:
```diff
@@ -74,7 +74,15 @@
      "shell.execute_reply": "2021-05-19T05:56:35.991124Z"
```
<p>Use f-strings here please</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='5'/>

pgbarletta replied:
<p>Fixed.</p>

IAlibay commented on doc/source/examples/analysis/alignment_and_rms/aligning_trajectory_to_frame.ipynb:
```diff
@@ -74,7 +74,15 @@
      "shell.execute_reply": "2021-05-19T05:56:35.991124Z"
```
<p>As above, f-strings</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='11'/>

pgbarletta replied:
<p>Fixed.</p>

IAlibay commented on doc/source/examples/analysis/alignment_and_rms/aligning_trajectory_to_frame.ipynb:
```diff
@@ -74,7 +74,15 @@
      "shell.execute_reply": "2021-05-19T05:56:35.991124Z"
```
<p>As above f-strings</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='13'/>

pgbarletta replied:
<p>Fixed.</p>

IAlibay commented on doc/source/examples/analysis/alignment_and_rms/aligning_trajectory_to_frame.ipynb:
```diff
@@ -74,7 +74,15 @@
      "shell.execute_reply": "2021-05-19T05:56:35.991124Z"
```
<p>as above f-strings</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='15'/>

pgbarletta replied:
<p>Fixed.</p>

IAlibay commented on doc/source/examples/analysis/alignment_and_rms/pairwise_rmsd.ipynb:
```diff
@@ -79,7 +79,15 @@
      "shell.execute_reply": "2021-05-19T05:56:39.153222Z"
```
<pre class="ql-syntax" spellcheck="false">Line #1.    plt.imshow(matrix.results.dist_matrix, cmap='viridis')
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='14'/>

pgbarletta replied:
<p>Fixed.</p>

IAlibay commented on doc/source/examples/analysis/alignment_and_rms/rmsd.ipynb:
```diff
@@ -87,7 +87,15 @@
      "shell.execute_reply": "2021-05-19T05:56:45.612777Z"
```
<pre class="ql-syntax" spellcheck="false">Line #1.    df = pd.DataFrame(R.results.rmsd, 
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='17'/>

pgbarletta replied:
<p>Fixed.</p>

IAlibay commented:
Yikes, ReviewNB makes more noise than I remember. I'm going through it in chunks (sorry!), but here is the first round of comments.

IAlibay commented:
The simtk warnings seem to be related to mda.Universe, but we removed them a long time ago (I want to say 2.2.0 or maybe even 2.1.0), what version of MDA was used to run these notebooks?

IAlibay commented on environment.yml:
```diff
@@ -36,11 +36,19 @@ dependencies:
   - threadpoolctl
-  - tabulate
-  - widgetsnbextension
+  - tabulate==0.8.7
```
Are these pins necessary? If so we need to work out why specifically and make it clear that we will undo them ASAP once the necessary conditions to remove the pins is possible (i.e. issues outlining the problem).

pgbarletta commented:
> The simtk warnings seem to be related to mda.Universe, but we removed them a long time ago (I want to say 2.2.0 or maybe even 2.1.0), what version of MDA was used to run these notebooks?

Yes, MDA has a `try/except` to avoid this. They actually come from parmed.

IAlibay commented:
> Is there an easy way to find out the minimum required version of MDA for each notebook or can I set all notebooks minimum version to 2.0.0? With the usage of the results attribute that's the case for most of them anyway.

There are no API breaking changes within versions of 2.x, so we can safely assume that unless something was added recently (say the pyedr work, or #213), then everything will be fine with 2.0.0.

> Can we speed up notebooks from trajectory_similarity? Perhaps getting some shorter trajectories?

Can just using frame skipping on calling `run` be the solution here? (adding new trajectories in MDAnalysisTests is a bit of a faff).

> Do we want to test HOLE as well?

In my opinion yes, at the very minimum the outputs of the notebook shouldn't have error traces.

> Can we lighten up volumetric/density_analysis.ipynb so we can check it as well?

This is already a ten frame trajectory so I don't know if we can cut it down much further, how would you recommend ligthening it up?

IAlibay commented:
> > The simtk warnings seem to be related to mda.Universe, but we removed them a long time ago (I want to say 2.2.0 or maybe even 2.1.0), what version of MDA was used to run these notebooks?
> 
> Yes, MDA has a `try/except` to avoid this. They actually come from parmed.

I can't reproduce the warnings you're getting though, they aren't occuring because parmed is being used (I have parmed installed locally in my mda dev environment too and it's not complaining if I trying to load a universe with PSD, DCD for example).

IAlibay commented on doc/source/examples/analysis/custom_parallel_analysis.ipynb:
```diff
@@ -71,12 +71,14 @@
   {
```
<p>Image seems to be missing? (or maybe ReviewNB doesn't render it properly anymore?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='51'/>

IAlibay replied:
<p>Just pinging this again, there doesn't seem to be any output displayed for jobs.visualize().</p>

IAlibay commented on doc/source/examples/analysis/custom_parallel_analysis.ipynb:
```diff
@@ -71,12 +71,14 @@
   {
```
<p>This seems to be a vastly different image, do we know why?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='55'/>

IAlibay replied:
<p>Pinging this again, the radius of gyration output here doesn't seem right, these shouldn't be straight lines?</p>

IAlibay commented on doc/source/examples/analysis/distances_and_contacts/distances_between_atomgroups.ipynb:
```diff
@@ -76,7 +76,17 @@
      "shell.execute_reply": "2021-05-19T05:57:16.889497Z"
```
<p>Add a filterwarning for this too?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='4'/>

pgbarletta replied:
<p>Fixed.</p>

IAlibay commented on doc/source/examples/analysis/distances_and_contacts/distances_between_selections.ipynb:
```diff
@@ -77,7 +77,17 @@
      "shell.execute_reply": "2021-05-19T05:57:19.319052Z"
```
<p>As above</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='4'/>

pgbarletta replied:
<p>Fixed.</p>

IAlibay commented on doc/source/examples/analysis/hydrogen_bonds/hbonds.ipynb:
```diff
@@ -72,7 +72,15 @@
    "cell_type": "code",
```
<pre class="ql-syntax" spellcheck="false">Line #1.    for frame, donor_ix, *_ in hbonds.results.hbonds:
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='44'/>

pgbarletta replied:
<p>Fixed.</p>

IAlibay commented on doc/source/examples/analysis/polymers_and_membranes/hole2.ipynb:
```diff
@@ -33,7 +33,7 @@
   },
```
<p>you'll need to add installing hole to CI, see: https://github.com/MDAnalysis/mdanalysis/blob/89d19d55349772be44a3e08eab8e54f586a3fd97/.github/actions/build-src/action.yaml#L39-L50</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='4'/>

IAlibay commented on doc/source/examples/analysis/reduced_dimensions/pca.ipynb:
```diff
@@ -37,7 +37,7 @@
     {
```
<p>All these <code>last executed</code> bits need to be updated or removed. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='0'/>

IAlibay commented on doc/source/examples/analysis/structure/average_rdf.ipynb:
```diff
@@ -74,8 +74,7 @@
      "name": "stderr",
```
<pre class="ql-syntax" spellcheck="false">Line #1.    plt.plot(irdf2.results.bins, irdf2.results.rdf)
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='16'/>

pgbarletta replied:
<p>Fixed.</p>

IAlibay commented on doc/source/examples/analysis/structure/elastic_network.ipynb:
```diff
@@ -77,7 +77,15 @@
      "shell.execute_reply": "2021-05-19T05:58:11.919074Z"
```
<p>I'm not seeing any changes on how this is being run, do we have a reason why the results here are drastically different? (did the core library change in some way?)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='9'/>

pgbarletta replied:
<p>good catch. It's just that <code>results</code> has a different meaning now, so that length is the number of keys in that dictionary (eigenvalues + eigenvectors + times).</p><p>Fixed. </p>

IAlibay commented on doc/source/examples/analysis/structure/site_specific_rdf.ipynb:
```diff
@@ -75,8 +75,7 @@
      "name": "stderr",
```
<pre class="ql-syntax" spellcheck="false">Line #1.    ca61_h2o_571 = ss_rdf.results.rdf[0][1][570]
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='13'/>

pgbarletta replied:
<p>Fixed.</p>

IAlibay commented on doc/source/examples/analysis/structure/site_specific_rdf.ipynb:
```diff
@@ -75,8 +75,7 @@
      "name": "stderr",
```
<pre class="ql-syntax" spellcheck="false">Line #1.    plt.plot(ss_rdf.results.bins, ca61_h2o_571)
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='14'/>

pgbarletta replied:
<p>Fixed.</p>

IAlibay commented on doc/source/examples/analysis/volumetric/linear_density.ipynb:
```diff
@@ -68,7 +68,15 @@
      "shell.execute_reply": "2021-05-19T06:03:14.365692Z"
```
<pre class="ql-syntax" spellcheck="false">Line #1.    plt.plot(np.linspace(0, 50, 200), density.results['x']['pos'])
</pre><p>Switch to the <code>results_density</code> alternative? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

IAlibay commented on doc/source/examples/constructing_universe.ipynb:
```diff
@@ -37,7 +37,7 @@
     {
```
<p>No output showing?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='49'/>

IAlibay replied:
<p>re-pinging this, is the lack of output here expected? (or is it ReviewNB being silly?)</p>

IAlibay commented on doc/source/examples/other/parmed_sim.ipynb:
```diff
@@ -31,7 +31,15 @@
      "start_time": "2020-12-27T11:14:14.063236Z"
```
<p>probably will want a filterwarning for this</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='37'/>

pgbarletta replied:
<p>Fixed.</p>

IAlibay commented:
> > > The simtk warnings seem to be related to mda.Universe, but we removed them a long time ago (I want to say 2.2.0 or maybe even 2.1.0), what version of MDA was used to run these notebooks?
> > 
> > 
> > Yes, MDA has a `try/except` to avoid this. They actually come from parmed.
> 
> I can't reproduce the warnings you're getting though, they aren't occuring because parmed is being used (I have parmed installed locally in my mda dev environment too and it's not complaining if I trying to load a universe with PSD, DCD for example).

Aha! I can now reproduce it, looks like I hadn't conda installed it. Could we just switch to using the `master` branch of parmed for now (and raise an issue to revert once the next release of parmed is done)? The namespace changes were done for parmed, but there's just not been a release since August 2021 (I'll ping some folks to see if there's a way to help with releases here).

My viewpoint here being that if we use the fixed version for now, we can avoid all the warning messages for now (given we're unlikely to re-run these notebooks regularly enough)?

lilyminium commented on doc/source/examples/quickstart.ipynb:
```diff
@@ -1,28 +1,29 @@
 {
```
<p>I wonder why you're getting this simtk deprecation warning? Are you somehow importing parmed as well?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='2'/>

pgbarletta replied:
<p>Fixed.</p>

lilyminium commented on doc/source/examples/quickstart.ipynb:
```diff
@@ -1,28 +1,29 @@
 {
```
<p>Hmm, it might be a good idea to sort this output so it doesn't change over each iteration?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='40'/>

pgbarletta replied:
<p>Fixed.</p>

lilyminium commented on doc/source/examples/quickstart.ipynb:
```diff
@@ -1,28 +1,29 @@
 {
```
<p>f-strings, please :)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='51'/>

pgbarletta replied:
<p>Fixed.</p>

lilyminium commented on doc/source/examples/quickstart.ipynb:
```diff
@@ -1,28 +1,29 @@
 {
```
<p>As above, f-strings</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='54'/>

pgbarletta replied:
<p>Fixed.</p>

lilyminium commented on doc/source/examples/quickstart.ipynb:
```diff
@@ -1,28 +1,29 @@
 {
```
<p>Would it be a good idea to append with <code>;</code> to suppress the memory output? Would the picture still render? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='70'/>

pgbarletta replied:
<p>works! It doesn't make the tests fail, but it does give a nicer output.</p>

lilyminium commented on doc/source/examples/transformations/center_protein_in_box.ipynb:
```diff
@@ -41,7 +41,7 @@
     {
```
<pre class="ql-syntax" spellcheck="false">Line #2.        protein_center = protein.center_of_mass(wrap=True)
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='13'/>

pgbarletta replied:
<p>thanks, missed that. Fixed.</p>

lilyminium commented on doc/source/examples/transformations/center_protein_in_box.ipynb:
```diff
@@ -41,7 +41,7 @@
     {
```
<p>Whoops. Raised https://github.com/MDAnalysis/mdanalysis/issues/3909</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='18'/>

pgbarletta replied:
<p>thanks for fixing it!</p>

lilyminium commented on doc/source/examples/constructing_universe.ipynb:
```diff
@@ -37,7 +37,7 @@
     {
```
<p>This simtk warning is also weird for an mda.Universe / nglview call :/</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='30'/>

IAlibay replied:
https://github.com/MDAnalysis/mdanalysis/blob/cef7d3fc566ec313a639a92506f81420da7c9358/package/MDAnalysis/__init__.py#L211

Looks like it's a _get_readers thing that loads up all the available converters :/

lilyminium commented:
> Aha! I can now reproduce it, looks like I hadn't conda installed it. Could we just switch to using the master branch of parmed for now (and raise an issue to revert once the next release of parmed is done)? The namespace changes were done for parmed, but there's just not been a release since August 2021 (I'll ping some folks to see if there's a way to help with releases here).

My strong preference would be to switch to a specific commit of parmed, rather than a `master` branch that may be continuously updated with possibly breaking changes. (Edit: I've been burned by this before in environment files).

pgbarletta commented:
<p>Fixed.</p>

---

 <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

pgbarletta commented:
<p>Fixed.</p>

---

 <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

pgbarletta commented:
<p>Personally, I'd keep it this way since static is what we want, right?</p>

---

 <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

pgbarletta commented:
Sorry for the delay, but this cell has been [hanging](https://github.com/pgbarletta/UserGuide/actions/runs/3435845515/jobs/5728682701) the CI and I couldn't find a way around it, so I decided to export the plot and add a markdown cell that loads it:
![image](https://user-images.githubusercontent.com/5693483/201114263-bd3a9833-ca86-44f2-be31-fff3e369b0d9.png)

Both the `p.show()` and `p.screenshot()` are commented out since they both hang the instance for some reason.

Other than that, the rest of the issues have been addressed:

- HOLE is downloaded and the hole.ipynb notebook is being tested
- parmed is now a pip dependency and is pinned to the last working commit I found. I'll open the issue to move this back to a conda release when this PR is merged.
- We now use a `pytest.ini` to simplify the tests workflow
- Miscellaneous requested fixes on the notebooks.

If anyone has any experience with an otherwise working cell freezing when ran on the CI pipeline, I'd appreciate any help.

IAlibay commented on .github/workflows/gh-ci-tests.yaml:
```diff
@@ -0,0 +1,58 @@
+          mamba list
+      - name: "test notebooks"
+        run: |
+            git clone https://github.com/osmart/hole2 --depth 1
+            cd hole2/src/
+            make
+            cd ../exe/
+            export PATH=$PATH:`pwd`
+            cd ../../tests
```
I'd split this into two sub-sections, it's better to isolate the hole2 build noise from the pytest noise so that you have less to filter through when something fails.

IAlibay commented on environment.yml:
```diff
@@ -36,11 +35,22 @@ dependencies:
-  - widgetsnbextension
+  - tabulate>=0.8.7
+  - widgetsnbextension==3.5.1
+  - Jinja2==2.11
```
Same comment as before, are these pins actually necessary? (we unpinned jinja2 a couple of PRs up), if so we need to work out why and what path is needed for us to unpin. Keeping things pinned is unfortunately too unssustainable and I'd like to make sure we can avoid it as much as possible.

pgbarletta replied:
Sorry, unpinned now.

IAlibay commented on .github/workflows/gh-ci-tests.yaml:
```diff
@@ -0,0 +1,58 @@
+          mamba list
+      - name: "test notebooks"
+        run: |
+            git clone https://github.com/osmart/hole2 --depth 1
```
I believe we don't pull from osmart's root branch because it's broken under some conditions. Please see/re-use from the MDA core CI as necessary: https://github.com/MDAnalysis/mdanalysis/blob/345d5d2d25a68df8ffac213fc3fe8bba2c9ec374/.github/actions/build-src/action.yaml#L39-L50

pgbarletta replied:
I copied that section but removed all the `FC` references. I guess that workflow sets the Fortran Compiler on some other place, but the instance already comes with `gfortran` and it works just fine.

IAlibay commented on .github/workflows/gh-ci-tests.yaml:
```diff
@@ -0,0 +1,58 @@
+        run: |
+          mamba info -a
+          mamba list
+      - name: "test notebooks"
```
```suggestion

      - name: "test notebooks"
```
Empty line for consistent formatting.

pgbarletta replied:
Fixed.

IAlibay commented on .github/workflows/gh-ci-tests.yaml:
```diff
@@ -0,0 +1,58 @@
+  tests:
+    runs-on: ubuntu-latest
+    name: "examples-tests"
+    strategy:
+      fail-fast: False
+      matrix:
+        os: ['ubuntu']
+        python-version: [3.8]
```
os isn't being used, just remove matrix altogether?

IAlibay reviewed (changes requested):
Not looked at the notebooks yet, please see some infrastructure comments.

pgbarletta commented:
By the way, I should add that I made a few changes on [`hole2.ipynb`](https://nbviewer.org/github/pgbarletta/UserGuide/blob/develop/doc/source/examples/analysis/polymers_and_membranes/hole2.ipynb) since now the notebook needs to run on the CI instance and the path to the `HOLE` binaries would get a bit uglier.

On the tests workflow I added the `hole2/bin` to the `PATH` variable of the runner ([here](https://github.com/pgbarletta/UserGuide/blob/7150994d1058e44a1f64acfbfb0a0e3ca9021ee5/.github/workflows/gh-ci-tests.yaml#L56)) so the notebook could run without explicitly pointing MDA to the `HOLE` binary:
![image](https://user-images.githubusercontent.com/5693483/201337743-9b4239b6-9b53-489d-a113-189a77220bcc.png)
I kept the original line commented out, instead of discarding it. It may not be very pretty, but this way the user still gets an example of both behaviours.
And I also guided the user through this choice:

![image](https://user-images.githubusercontent.com/5693483/201339634-516c32a4-519c-4c34-b044-803e3d73f24d.png)

IAlibay commented on .gitignore:
```diff
@@ -48,3 +48,5 @@ doc/source/examples/**/*.*
 doc/source/examples/**/.*.ipynb
 
+# ignore HOLE2 local install folder
+hole2
```
Ideally this shouldn't be done inside of `userguide` to begin with, but it's a small enough change that I guess it doesn't matter.

pgbarletta replied:
I was about to just ignore it on my `.git/info/exclude` but thought that other devs may run into it as well. I can take it out if you want

IAlibay commented on environment.yml:
```diff
@@ -24,23 +24,32 @@ dependencies:
       - sphinx-sitemap
+      - nbval>=0.9.6
+      - pyedr>=0.7
+      - git+https://github.com/ParmEd/ParmEd.git@7f837a0c0010299f24edce8240514c546944c458
```
Could you raise an issue to unpin this please? That way we can remember to do it.

pgbarletta replied:
Yes, I'd like to do it once is merged if you don't mind. Promise I won't forget :)

IAlibay reviewed (commented):
One non-blocking comment, I'll review the rest via ReviewNB

IAlibay commented on doc/source/examples/analysis/alignment_and_rms/aligning_trajectory.ipynb:
```diff
@@ -8,7 +8,7 @@
     "\n",
```
<p>Do we still need the "uncomment these lines" bit? (I don't see any commented lines in the diff anymore).</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='23'/>

pgbarletta replied:
oops. Fixed.

IAlibay commented on doc/source/examples/analysis/custom_parallel_analysis.ipynb:
```diff
@@ -10,9 +10,9 @@
     "\n",
```
<p>Some of these stale lock files seem to be related to existence of a directory for previous worker that hasn't cleaned up properly. Can this be avoided? I don't seem to be able to replicate these info messages locally.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='25'/>

pgbarletta replied:
my bad. Fixed.

IAlibay commented on doc/source/examples/analysis/custom_parallel_analysis.ipynb:
```diff
@@ -10,9 +10,9 @@
     "\n",
```
<p>setting n_jobs to cpu_count is going to lead to this changing between machines (since folks will have different machines). Would it be best to just hardcode it to a given value?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='47'/>

pgbarletta replied:
given that the notebook needs to run on the instance, 2 is the only real option.
I'd still leave this commented out
```
from multiprocessing import cpu_count
n_jobs = cpu_count()
```
since it's useful for the user.

pgbarletta commented:
Fixed now. 

### About `custom_parallel_analysis.ipynb`:

Due to [this](https://github.com/MDAnalysis/mdanalysis/issues/3931) issue I had to pin [these](https://github.com/pgbarletta/UserGuide/blob/b7db05dfa2a19be0c8ea7896de032ec0ec793afc/environment.yml#L7) Dask dependencies and specify `pytest`:
```
  - distributed==2021.1.0
  - dask==2021.1.0
  - python-graphviz
  - pytest
```

IAlibay commented:
> Fixed now. 
>
>### About `custom_parallel_analysis.ipynb`:
>
>Due to [this](https://github.com/MDAnalysis/mdanalysis/issues/3931) issue I had to pin [these](https://github.com/pgbarletta/UserGuide/blob/b7db05dfa2a19be0c8ea7896de032ec0ec793afc/environment.yml#L7) Dask dependencies and specify `pytest`:

As per other things, please do raise issues reminding us to unpin these as soon as possible.

IAlibay commented on .github/workflows/gh-ci-tests.yaml:
```diff
@@ -0,0 +1,61 @@
+          mamba list
+
+      - name: build_hole
+        shell: bash -l {0}
```
```suggestion
```
_should_ be already defined by default hopefully

IAlibay reviewed (changes requested):
Just the one thing at this level, I'll do the rest using reviewNB.

(I'll directly push the suggestion so we can check CI is working directly).

IAlibay commented on doc/source/examples/analysis/structure/elastic_network.ipynb:
```diff
@@ -8,9 +8,9 @@
     "\n",
```
<p>This plot is completely different now, is this expected?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='13'/>

pgbarletta replied:
<p>Fixed!</p>

IAlibay commented on doc/source/examples/analysis/structure/elastic_network.ipynb:
```diff
@@ -8,9 +8,9 @@
     "\n",
```
<p>As above, this doesn't seem to show much anymore?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='15'/>

pgbarletta replied:
<p>Fixed.</p>

IAlibay commented on doc/source/examples/analysis/structure/elastic_network.ipynb:
```diff
@@ -8,9 +8,9 @@
     "\n",
```
<p>As above</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='20'/>

pgbarletta replied:
<p>Fixed.</p>

IAlibay commented on doc/source/examples/analysis/structure/elastic_network.ipynb:
```diff
@@ -8,9 +8,9 @@
     "\n",
```
<p>As above</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

pgbarletta replied:
<p>Fixed.</p>

IAlibay commented on doc/source/examples/analysis/structure/site_specific_rdf.ipynb:
```diff
@@ -8,9 +8,9 @@
     "\n",
```
<p>Just checking - are we expecting the changes in output seen here compared to before?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='14'/>

pgbarletta replied:
<p>thanks for catching it. This was due to the change from 0 to 1 indexing with the TPR topologies. Fixed now.</p>

IAlibay commented on doc/source/examples/constructing_universe.ipynb:
```diff
@@ -8,7 +8,7 @@
     "\n",
```
<p>The outputs here are going to change whenever we update the mda / nglview versions, is this going to mess with CI?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='1'/>

pgbarletta replied:
<p>There's a sanitization regex rule on <code>sanitize_output.cfg</code> for these messages:</p><p><code></code><code></p><p>[Version] </p><p>regex: (Using .* version.*) </p><p>replace: VERSION</p><p> </code><code></code></p><p> </p>

IAlibay commented on doc/source/examples/quickstart.ipynb:
```diff
@@ -1,28 +1,29 @@
 {
```
<p>As above re: MDA versions changing over time</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='2'/>

IAlibay commented on doc/source/examples/quickstart.ipynb:
```diff
@@ -1,28 +1,29 @@
 {
```
<p>Might as well update 1.0.0 here?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='3'/>

pgbarletta replied:
<p>Fixed.</p>

IAlibay commented on doc/source/examples/quickstart.ipynb:
```diff
@@ -1,28 +1,29 @@
 {
```
<p>Do we want to filter warnings for this?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='8'/>

pgbarletta replied:
<p>Fixed.</p>

pgbarletta commented:
<p>Yes, I explained why I chose to keep it [here](https://github.com/MDAnalysis/UserGuide/pull/241#issuecomment-1311635374)</p>

---

 <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

IAlibay commented:
<p>I understand that you've added it to PATH, but it's no longer actually installed under <code>~/hole2/exe/hole</code> right? (hence the need to have it under $PATH?).</p><p> </p><p>I full agree with the commented out inputs pointing to a given location, but at least as far as I understand things currently, wouldn't it be better to a) make it clear why they are commented out, b) say that in this case we've use the PATH option but we've added the commented out entries in case they want to do it via path?</p>

---

 <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

pgbarletta commented:
<p>Added a comment to clarify that we're using the PATH variable. I'll commit it after fixing the new issues that've shown up.</p>

---

 <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

pgbarletta commented:
<p>Added.</p>

---

 <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

pgbarletta commented:
Alright, tests are passing. Once [this](https://github.com/MDAnalysis/mdanalysis/pull/3951) is merged I'll re-run the notebooks below so they look nicer (without the warning from Bio) and squash everything into 1 commit.
```
doc/source/examples/analysis/alignment_and_rms/aligning_structure_to_another.ipynb
doc/source/examples/analysis/alignment_and_rms/aligning_trajectory_to_frame.ipynb
doc/source/examples/analysis/alignment_and_rms/pairwise_rmsd.ipynb
doc/source/examples/analysis/alignment_and_rms/rmsd.ipynb
doc/source/examples/analysis/custom_parallel_analysis.ipynb
doc/source/examples/analysis/custom_trajectory_analysis.ipynb
doc/source/examples/analysis/distances_and_contacts/contacts_custom.ipynb
doc/source/examples/analysis/distances_and_contacts/contacts_native_fraction.ipynb
doc/source/examples/analysis/distances_and_contacts/contacts_q1q2.ipynb
doc/source/examples/analysis/distances_and_contacts/contacts_within_cutoff.ipynb
doc/source/examples/analysis/hydrogen_bonds/hbonds-lifetimes.ipynb
doc/source/examples/analysis/hydrogen_bonds/hbonds-selections.ipynb
doc/source/examples/analysis/hydrogen_bonds/hbonds.ipynb
doc/source/examples/analysis/polymers_and_membranes/hole2.ipynb
doc/source/examples/analysis/reduced_dimensions/diffusion_map.ipynb
doc/source/examples/analysis/structure/elastic_network.ipynb
doc/source/examples/analysis/structure/helanal.ipynb
doc/source/examples/analysis/structure/site_specific_rdf.ipynb
doc/source/examples/analysis/trajectory_similarity/clustering_ensemble_similarity.ipynb
doc/source/examples/analysis/trajectory_similarity/convergence.ipynb
doc/source/examples/analysis/trajectory_similarity/dimension_reduction_ensemble_similarity.ipynb
doc/source/examples/analysis/trajectory_similarity/harmonic_ensemble_similarity.ipynb
doc/source/examples/analysis/volumetric/linear_density.ipynb
doc/source/examples/constructing_universe.ipynb
doc/source/examples/quickstart.ipynb
```

orbeckst commented:
@pgbarletta , PR https://github.com/MDAnalysis/mdanalysis/pull/3951 was merged.

pgbarletta commented:
Thanks @orbeckst.

The warning from Bio is gone and I changed the **Last updated** header to Dec 2022 on all noteboks. Also removed the **Last executed** header since now all notebooks are continuously tested.

IAlibay commented on doc/source/examples/analysis/custom_parallel_analysis.ipynb:
```diff
@@ -1,6 +1,7 @@
 {
```
<p>Here and everywhere else, do we need to filterwarning for this DCD reader deprecation?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='6'/>

IAlibay replied:
Constantly running after warnings seems unsustainable, I'd be happy if the answer here is "it won't mess with CI, let's just keep it in because that's how it'll look to users" and we do nothing further on this one.

pgbarletta replied:
<p>Currently I'm sanitizing <code>UserWarning</code> and  <code>DeprecationWarning</code>, so all the warning text get's replaced by something  like a <strong>USERWARNING</strong> when nbval evaluates it. This way, the CI doesn't fail due to repeated warnings, but it will still fail when a new warning shows up since they create the <code>stderr</code> attribute on the cell and if that wasn't present before, nbval will fail the CI. At the same time, we can't ignore the <code>stderr </code> .</p><p> </p><p>I don't know how often new warnings show up, either from MDA or some of its dependencies, but if it's a weekly thing, then yes, perhaps we should add a:</p><pre class="ql-syntax" spellcheck="false">import warnings
# suppress some MDAnalysis warnings
warnings.filterwarnings('ignore') 
</pre><p>on each notebook. Cells will still be tested and if they fail or their <code>stdout</code> changes nbval will catch them still.</p>

IAlibay replied:
<p>I'm going to leave this to a future problem and we can see how often CI fails.</p>

IAlibay commented on doc/source/examples/analysis/hydrogen_bonds/hbonds-lifetimes.ipynb:
```diff
@@ -8,7 +8,7 @@
     "\n",
```
<p>oof this is a lot noisier than I expected.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='29'/>

IAlibay replied:
Note: this is just a comment, we don't need to do anything, but we should be more careful upstream :/

pgbarletta replied:
<p>Added a warning filter to ignore it.</p>

IAlibay commented on doc/source/examples/constructing_universe.ipynb:
```diff
@@ -1,16 +1,15 @@
 {
```
<p>Re-pinging this one as a separate comment, I'm not seeing any output here for some reason.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='49'/>

pgbarletta replied:
<p>Yes, it's a <code>None</code> because dimensions are set afterwards. I added a <code>print(u.dimensions)</code> to make it more patent. </p>

pgbarletta commented:
<p>oops, sorry. Fixed.</p>

---

 <a href='https://app.reviewnb.com/MDAnalysis/UserGuide/pull/241/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

pgbarletta committed:
Add nbval testing for all notebooks.

 - Fix notebooks so tests pass
 - Add necessary dependencies

IAlibay reviewed (approved):
lgtm! @lilyminium did you want to have another look at this before I merge?

IAlibay commented:
I got an offline lgtm from @lilyminium so we can go ahead with the merge. Thanks @pgbarletta !
